helix-term/src/ui/completion.rs

```diff
@@ -19,6 +19,12 @@ use helix_lsp::{lsp, util, OffsetEncoding};
 
 impl menu::Item for CompletionItem {
     type Data = ();
+
+    #[inline]
+    fn group(&self, _data: &Self::Data) -> usize {
+        self.language_server_id
+    }
```

Comment:
I can't think of other places we would want to re-use this `group` concept. Ideally I'd like to see this built in to the `sort_text` function: something like `fn sort_key(&self, data: &Self::Data) -> impl Ord` and then `CompletionItem`'s `Item` implementation would contain the logic for using `self.language_server_id`. That depends on either https://github.com/rust-lang/rust/issues/91611 to return an `impl Ord`, or https://github.com/rust-lang/rust/issues/29661 so that we wouldn't have to explicitly implement `sort_key` in every Item:

```rust
pub trait Item {
    type Data;
    type SortKey: Ord = Cow<str>;

    // ...

    fn sort_key(&self, data: &Self::Data) -> Self::SortKey {
        self.filter_text()
    }

    // ...
}
```

Maybe there's a clever way around the language features here that I'm not seeing though?

There was a similar thread of this idea in https://github.com/helix-editor/helix/pull/2705#discussion_r927234858.

Reply:
- removed group concept.
- introduced `Item.score` function to control score and sorting